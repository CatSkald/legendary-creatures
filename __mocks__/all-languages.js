import languages from "../src/i18n/languages";

exports.allLanguages = Object.entries(languages).map(
  ([lang, langProps]) => langProps,
);
