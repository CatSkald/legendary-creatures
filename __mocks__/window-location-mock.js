exports.mockWindowLocation = () => {
  const originalWindowLocation = window.location;

  beforeEach(() => {
    delete window.location;

    // mock window.location
    window.location = Object.defineProperties(
      {},
      {
        ...Object.getOwnPropertyDescriptors(originalWindowLocation),
        assign: {
          configurable: true,
          value: jest.fn(),
        },
      },
    );
  });

  afterEach(() => {
    // restore original `jsdom` location
    window.location = originalWindowLocation;
  });
};
