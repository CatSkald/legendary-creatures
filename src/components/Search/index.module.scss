.searchButton {
  text-align: center;
  text-decoration: none;
  font-weight: bold;
  margin: var(--space-sm);
  padding: var(--space-sm);
  border-radius: var(--border-radius);
  border: solid;
  border-width: var(--border-width);
  border-color: var(--border-color);
  background-color: var(--card-bg);
  color: var(--card-link);
  &:hover {
    color: var(--card-link-hover);
    border-color: var(--content-link-hover);
  }
}

.searchInput {
  margin: var(--space-sm);
  border-radius: var(--border-radius);
}
