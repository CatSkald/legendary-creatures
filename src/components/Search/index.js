import { searchInput, searchButton } from "./index.module.scss";

import React from "react";
import PropTypes from "prop-types";

import useTranslations from "../../i18n/translations/useTranslations";

const Search = ({ searchEngine, queryTemplate }) => {
  const createSearchQuery = (term) => `${queryTemplate}${term}`;
  const translations = useTranslations();

  const [searchQuery, onChangeSearchQuery] = React.useState(
    createSearchQuery(""),
  );

  const handleOnKeyDown = (e) => {
    if (searchQuery && e.key === "Enter") {
      window.location.assign(searchQuery);
      return false;
    }
  };
  return (
    <>
      <label>
        {translations.SearchUsing} {searchEngine}:
        <input
          className={searchInput}
          type="search"
          name="q"
          autoFocus
          onChange={(e) =>
            onChangeSearchQuery(createSearchQuery(e.target.value))
          }
          onKeyPress={handleOnKeyDown}
        />
      </label>
      <a className={searchButton} href={searchQuery}>
        {translations.Search}
      </a>
    </>
  );
};

Search.propTypes = {
  searchEngine: PropTypes.string.isRequired,
  queryTemplate: PropTypes.string.isRequired,
};

export default Search;
