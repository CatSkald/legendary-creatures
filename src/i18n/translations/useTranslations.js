import React from "react";
import { useStaticQuery, graphql } from "gatsby";
import { LocaleContext } from "../../components/Layout";

function useTranslations() {
  const { language } = React.useContext(LocaleContext);
  const { rawData } = useStaticQuery(queryTranslations);

  const translationsForCurrentLocale = rawData.edges
    .filter(item => item.node.name === language.code)
    .map(item => item.node.translations)[0];

  return translationsForCurrentLocale;
}

export default useTranslations;

const queryTranslations = graphql`
  query queryTranslations {
    rawData: allFile(filter: { sourceInstanceName: { eq: "translations" } }) {
      edges {
        node {
          name
          translations: childTranslationsYaml {
            Description
            Title
            ShortTitle
            NotFound404Header
            NotFound404Content
            footerMoto
            by
            for
            fromPerson
            fromPlace
            of
            Categories
            View
            all
            orSearchBy
            tags
            All
            Home
            Latest
            Next
            Prev
            Tags
            plural
            Number
            Origin
            Habitat
            CreaturesPageTitle
            TagsPageTitle
            origin
            categories
            number
            habitat
          }
        }
      }
    }
  }
`;
