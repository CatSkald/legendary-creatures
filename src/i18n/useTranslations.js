import React from "react";
import { useStaticQuery, graphql } from "gatsby";
import { LocaleContext } from "../components/Layout";

function useTranslations() {
  const { locale } = React.useContext(LocaleContext);
  const { rawData } = useStaticQuery(queryTranslations);

  const translationsForCurrentLocale = rawData.edges
    .filter(item => item.node.name === locale)
    .map(item => item.node.translations)[0];

  return translationsForCurrentLocale;
}

export default useTranslations;

const queryTranslations = graphql`
  query queryTranslations {
    rawData: allFile(
      filter: { sourceInstanceName: { eq: "translations" } }
    ) {
      edges {
        node {
          name
          translations: childTranslationsYaml {
            Description
            Title
            NotFound404Header
            NotFound404Content
            footerMoto
            Icon
            by
            for
            fromPerson
            fromPlace
            of
            using
            Categories
            All
            Home
            Latest
            Next
            Prev
            Tags
            plural
            Number
            Origin
            Habitat
          }
        }
      }
    }
  }
`;
