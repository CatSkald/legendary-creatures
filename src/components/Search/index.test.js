import React from "react";
import { fireEvent, render } from "@testing-library/react";

import Search from "../Search";
import useTranslations from "../../i18n/translations/useTranslations";

import { mockWindowLocation } from "../../../__mocks__/window-location-mock";

mockWindowLocation();
jest.mock("../../i18n/translations/useTranslations");

useTranslations.mockImplementation(() => ({
  useTranslations: () => ({
    SearchUsing: () => "Search using",
    Search: () => "Search",
  }),
}));

describe("Search", () => {
  const queryTemplate = "https://localhost/?q=";

  test("renders correctly", () => {
    const { container } = render(
      <Search searchEngine="Search Me" queryTemplate={queryTemplate} />,
    );
    expect(container).toMatchSnapshot();
  });

  test("update link with entered search term", () => {
    const searchText = "text";
    const { container } = render(
      <Search searchEngine="Search Me" queryTemplate={queryTemplate} />,
    );
    const input = container.querySelector("input[type=search");
    fireEvent.change(input, { target: { value: searchText } });
    const link = container.querySelector("a");
    expect(link.href).toBe(`${queryTemplate}${searchText}`);
  });

  test("pressing Enter clicks link", () => {
    const searchText = "text";
    const { container } = render(
      <Search searchEngine="Search Me" queryTemplate={queryTemplate} />,
    );
    const input = container.querySelector("input[type=search");
    fireEvent.change(input, { target: { value: searchText } });
    fireEvent.keyPress(input, { key: "Enter", code: "Enter", charCode: 13 });

    expect(window.location.assign).toHaveBeenCalledTimes(1);
    expect(window.location.assign).toHaveBeenCalledWith(
      `${queryTemplate}${searchText}`,
    );
  });
});
