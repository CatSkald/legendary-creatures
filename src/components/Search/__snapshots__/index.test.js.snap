// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div>
  <label>
     
    Search Me
    :
    <input
      class="searchInput"
      name="q"
      type="search"
    />
  </label>
  <a
    class="searchButton"
    href="https://localhost/?q="
  />
</div>
`;
